rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,641 @@
+concurrency issues during the rolling. We need to find a way to have this rolling stage clustered-proof.
+
+#### Options we considered:
+
+1. have the rolling file appenders coordinate themselves when rolling
+
+By using a broadcast message based mutex mechanism, the appenders could acquire a ‘lock’ to roll a specific file, and
+notify other workers when the rolling is complete (quite similar to what we want to do with SO migration for example).
+
+An alternative to this option would be to only have the main worker handle the rolling logic. We will lose control
+on the exact size the file is when rolling, as we would need to wait until the main worker receives a log message
+for the rolling appender before the rolling is effectively performed. The upside would be that it reduces the inter-workers
+communication to a notification from the main worker to the others once the rolling is done for them to reopen their
+file handler.
+
+2. have the coordinator process perform the rolling
+
+Another option would be to have the coordinator perform the rotation instead. When a rolling is required, the appender 
+would send a message to the coordinator, which would perform the rolling and notify the workers once the operation is complete. 
+
+Note that this option is even more complicated than the previous one, as it forces to move the rolling implementation 
+outside of the appender, without any significant upsides identified.
+
+3. centralize the logging system in the coordinator
+
+We could go further, and change the way the logging system works in clustering mode by having the coordinator centralize 
+the logging system. The worker’s logger implementation would just send messages to the coordinator. If this may be a 
+correct design, the main downside is that the logging implementation would be totally different in cluster and 
+non cluster mode, and seems to be way more work that the other options.
+
+#### Our recommended approach:
+Even though it's more complex, we feel that centralizing the logging system in the coordinator is the right move here,
+as it will also solve for how to enable the coordinator to log its own messages.
+
+### 6.1.3 The status API
+
+In clustering mode, the workers will all have an individual status. One could have a connectivity issue with ES 
+while the other ones are green. Hitting the `/status` endpoint will reach a random (and different each time) worker, 
+meaning that it would not be possible to know the status of the cluster as a whole.
+
+We will need to add some centralized status state in the coordinator. Also, as the `/status` endpoint cannot be served 
+from the coordinator, we will also need to have the workers retrieve the global status from the coordinator to serve 
+the status endpoint.
+
+We may also want to have the `/status` endpoint display each individual worker status in addition to the 
+global status, which may be breaking change in the `/status` API response format.
```

Comment:
I think we can achieve both without breaking the API by having an overall status that shows the highest-severity status level for each component and the per-worker status as an extension of the existing response shape.

Reply:
Ah, I missed updating this section, thanks